### ISSUE TITLE
TyLayout::field_type ICE on 1.34.0-nightly 865cb7010 2019-02-10

### ISSUE THREAD

**Issue by lukaslueg** [rust-lang/rust#58610]
Running on `cargo 1.34.0-nightly (865cb7010 2019-02-10)` the travis job failed with a panic on [this commit](https://github.com/lukaslueg/rust-headless-chrome/commit/4fc1f843f6230e1083f357f3bf32e3e87e512b9d)

```
   Compiling headless_chrome v0.1.3 (/home/travis/build/lukaslueg/rust-headless-chrome)
error: internal compiler error: src/librustc/ty/layout.rs:1656: TyLayout::field_type(TyLayout { ty: u64, details: LayoutDetails { variants: Single { index: 0 }, fields: Union(0), abi: Scalar(Scalar { value: Int(I64, false), valid_range: 0..=18446744073709551615 }), align: AbiAndPrefAlign { abi: Align { pow2: 3 }, pref: Align { pow2: 3 } }, size: Size { raw: 8 } } }): not applicable

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:620:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:70
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:58
             at src/libstd/panicking.rs:200
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:215
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:482
   6: std::panicking::begin_panic
   7: rustc_errors::Handler::bug
   8: rustc::util::bug::opt_span_bug_fmt::{{closure}}
   9: rustc::ty::context::tls::with_opt::{{closure}}
  10: rustc::ty::context::tls::with_context_opt
  11: rustc::ty::context::tls::with_opt
  12: rustc::util::bug::opt_span_bug_fmt
  13: rustc::util::bug::bug_fmt
  14: <rustc_target::abi::TyLayout<'a, Ty>>::field
  15: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::operand_field
  16: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::eval_place_to_op
  17: rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::run
  18: rustc_mir::const_eval::eval_body_using_ecx
  19: rustc_mir::const_eval::const_eval_raw_provider
  20: rustc::ty::query::__query_compute::const_eval_raw
  21: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::const_eval_raw<'tcx>>::compute
  22: rustc::dep_graph::graph::DepGraph::with_task_impl
  23: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  24: rustc_mir::const_eval::const_eval_provider
  25: rustc::ty::query::__query_compute::const_eval
  26: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::const_eval<'tcx>>::compute
  27: rustc::dep_graph::graph::DepGraph::with_task_impl
  28: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  29: rustc_mir::const_eval::const_eval_provider
  30: rustc::ty::query::__query_compute::const_eval
  31: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::const_eval<'tcx>>::compute
  32: rustc::dep_graph::graph::DepGraph::with_task_impl
  33: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  34: rustc_mir::monomorphize::collector::collect_items_rec
  35: rustc_mir::monomorphize::collector::collect_items_rec
  36: rustc_mir::monomorphize::collector::collect_items_rec
  37: rustc_mir::monomorphize::collector::collect_items_rec
  38: rustc_mir::monomorphize::collector::collect_items_rec
  39: rustc_mir::monomorphize::collector::collect_items_rec
  40: rustc_mir::monomorphize::collector::collect_crate_mono_items::{{closure}}
  41: rustc::util::common::time
  42: rustc_mir::monomorphize::collector::collect_crate_mono_items
  43: rustc::util::common::time
  44: rustc_mir::monomorphize::partitioning::collect_and_partition_mono_items
  45: rustc::ty::query::__query_compute::collect_and_partition_mono_items
  46: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::collect_and_partition_mono_items<'tcx>>::compute
  47: rustc::dep_graph::graph::DepGraph::with_task_impl
  48: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  49: rustc_codegen_ssa::base::codegen_crate
  50: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_utils::codegen_backend::CodegenBackend>::codegen_crate
  51: rustc::util::common::time
  52: rustc_driver::driver::phase_4_codegen
  53: rustc_driver::driver::compile_input::{{closure}}
  54: <std::thread::local::LocalKey<T>>::with
  55: rustc::ty::context::TyCtxt::create_and_enter
  56: rustc_driver::driver::compile_input
  57: rustc_driver::run_compiler_with_pool
  58: <scoped_tls::ScopedKey<T>>::set
  59: rustc_driver::run_compiler
  60: syntax::with_globals
  61: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:87
  62: <F as alloc::boxed::FnBox<A>>::call_box
  63: std::sys::unix::thread::Thread::new::thread_start
             at /rustc/146aa60f3484d8267e085e80611969f387eca068/src/liballoc/boxed.rs:749
             at src/libstd/sys_common/thread.rs:14
             at src/libstd/sys/unix/thread.rs:81
  64: start_thread
  65: clone
query stack during panic:
#0 [const_eval_raw] const-evaluating `browser::waiting_helpers::wait_for_mut`
#1 [const_eval] const-evaluating + checking `browser::waiting_helpers::wait_for_mut`
#2 [const_eval] const-evaluating + checking `browser::waiting_helpers::wait_for_mut`
#3 [collect_and_partition_mono_items] collect_and_partition_mono_items
end of query stack
error: aborting due to previous error
```

**Comment by jonas-schievink** [comment]
Works on stable and beta

**Comment by pnkfelix** [comment]
triage: P-high.

**Comment by pnkfelix** [comment]
It would be good to factor out a standalone test here. All I can currently tell is that this bug was exposed by lukaslueg/rust-headless-chrome@4fc1f843f6230e1083f357f3bf32e3e87e512b9d

**Comment by pnkfelix** [comment]
triage: assigning to @csmoe for initial investigation (e.g. factoring out standalone test)

**Comment by pnkfelix** [comment]
removing nomination label since this is now prioritized and assigned

**Comment by csmoe** [comment]
```rust
use std::time::Duration;
fn main() {
    let chrome_output_result = foo();
}
struct A {
    a: u64,
}
fn foo() -> Result<(), ()> {
    let a = A { a: 1u64 };
    wait_for_mut(|| Some(1i32), a)
}

fn wait_for_mut<F, i32>(mut predicate: F, a: A) -> Result<(), ()>
where
    F: FnMut() -> Option<i32>,
{
    loop {
        if Duration::new(0, 0) > Duration::from_millis(a.a) {
            return Err(().into());
        }
    }
}

```

**Comment by oli-obk** [comment]
I'm guessing this is another case of the promotion bug that will be fixed by https://github.com/rust-lang/rust/pull/58784

**Comment by lukaslueg** [comment]
It seems this was indeed fixed by #58784
